# Classification 2

We will look at the Naive Bayes classifier.

* Only slightly more complicated than nearest-neighbors in terms of parameter fitting
* Scales to large datasets easily
* Works best when features are _independent_

## Income Prediction

This data was extracted from a census bureau dataset. It contains several features about individuals: their age, working status, education level, marital status, etc. The goal is to predict if the individual makes more or less than 50K per year.

See [here](https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.names "Adult names") for details.

In [1]:
from pandas import Series, DataFrame
import pandas as pd
from patsy import dmatrices
%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [2]:
# We will ignore some silly warnings that pop up due to scikit-learn
import warnings
warnings.filterwarnings('ignore')

In [3]:
names = ['age', 'workclass', 'fnlwgt', 'education', 'education_num', 'marital_status', \
         'occupation', 'relationship', 'race', 'sex', 'capital_gain', 'capital_loss', \
         'hours_per_week', 'native_country', 'income_band']
df = pd.read_csv('Classification_2_data/adult.data', sep=', ', names=names)
df[:5]

age         workclass  fnlwgt  education  education_num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital_status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital_gain  capital_loss  hours_per_week native_country income_band  
0          2174             0              40  United-States       <=50K  
1             0             0              13  United-States       <=50K  
2             0             0              40  United-States       <=50K  
3             0             0              40  United-States       <=50K  
4             0             0              40           Cuba       <=50K

In [4]:
df.describe()

age        fnlwgt  education_num  capital_gain  capital_loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours_per_week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

#### Create the target

In [5]:
df['target'] = 0.0
mask = (df['income_band'] == '>50K')
df['target'][mask] = 1.0

In [6]:
df['target'].value_counts()

0.0    24720
1.0     7841
Name: target, dtype: int64

### Discretize continuous values

While Naive Bayes can be run with continuous values, we will try it only with categorical features

* **workclass** is categorical; it takes values among
    * Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
* **age** is continuous.

For the continuous features, we will _discretize_ them into bins. Thankfully, pandas gives us a simple way to do so.

In [7]:
# NOTE: We also return the bins (retbins=True).
# We will use these bins later on the test set.
df['age_binned'], age_bins = pd.qcut(df['age'], 5, retbins=True)
print(df['age_binned'][:5])

0    (33.0, 41.0]
1    (41.0, 50.0]
2    (33.0, 41.0]
3    (50.0, 90.0]
4    (26.0, 33.0]
Name: age_binned, dtype: category
Categories (5, interval[float64, right]): [(16.999, 26.0] < (26.0, 33.0] < (33.0, 41.0] < (41.0, 50.0] < (50.0, 90.0]]


In [8]:
df['fnlwgt_binned'], fnlwgt_bins = pd.qcut(df['fnlwgt'], 5, retbins=True)
df['education_num_binned'], education_num_bins = pd.qcut(df['education_num'],
                                                         3, retbins=True)

Other seemingly continuous features are actually almost categorical. Consider capital_loss, for instance.

In [9]:
df['capital_loss'].value_counts()[:5]

0       31042
1902      202
1977      168
1887      159
1848       51
Name: capital_loss, dtype: int64

In [10]:
len(df)

32561

So, let us bin capital loss as either 0, or greater than 0.

In [11]:
df['capital_loss_binned'] = '0'
df['capital_loss_binned'][df['capital_loss'] > 0] = '>0'

In [12]:
df['capital_gain_binned'] = '0'
df['capital_gain_binned'][df['capital_gain'] > 0] = '>0'

Similarly for hours_per_week: less than 40, 40, and more than 40.

In [13]:
df['hours_binned'] = '40'
df['hours_binned'][df['hours_per_week'] < 40] = '<40'
df['hours_binned'][df['hours_per_week'] > 40] = '>40'

#### Create the design matrices

In [14]:
print(df.columns.values)

['age' 'workclass' 'fnlwgt' 'education' 'education_num' 'marital_status'
 'occupation' 'relationship' 'race' 'sex' 'capital_gain' 'capital_loss'
 'hours_per_week' 'native_country' 'income_band' 'target' 'age_binned'
 'fnlwgt_binned' 'education_num_binned' 'capital_loss_binned'
 'capital_gain_binned' 'hours_binned']


For Naive Bayes, we want all dummy variables, but unfortunately standard patsy doesn't necessarily give us that. So, just for the Naive Bayes classifier, we will use a pandas-specific way of dummy encoding.

In [15]:
# Make a list of all our categorical variables
categorical_columns = ['age_binned', 'workclass', 'fnlwgt_binned', 'education',
          'education_num_binned', 'marital_status', 'occupation', 'relationship',
          'race', 'sex', 'capital_gain_binned', 'capital_loss_binned', 'hours_binned']

In [16]:
# Use pandas to create dummy variables
df_dummies = pd.get_dummies(df[categorical_columns],
                            prefix=categorical_columns,
                            columns=categorical_columns)
dummy_column_names = df_dummies.columns.values
dummy_column_names[:10]

array(['age_binned_(16.999, 26.0]', 'age_binned_(26.0, 33.0]',
       'age_binned_(33.0, 41.0]', 'age_binned_(41.0, 50.0]',
       'age_binned_(50.0, 90.0]', 'workclass_?', 'workclass_Federal-gov',
       'workclass_Local-gov', 'workclass_Never-worked',
       'workclass_Private'], dtype=object)

In [17]:
# Concatenate all these new dummy columns into the old dataframe
df2 = pd.concat([df, df_dummies], axis=1)

Now, we use patsy to predict target using all the dummy variables.
* But we have dummy variables with strange names
* If we say `"target ~ age_binned_(16.999, 26.0]"`
    * patsy will get confused by the "(" and the space.

* We need to put the feature names in a special format for patsy
* `age_binned_(16.999, 26.0]` &rarr; `Q("age_binned_(16.999, 26.0]")`

In [18]:
formula = 'target ~ 0 + {}'.format(' + '.join(['Q("{}")'.format(x)\
                                               for x in dummy_column_names]))
print(formula)

target ~ 0 + Q("age_binned_(16.999, 26.0]") + Q("age_binned_(26.0, 33.0]") + Q("age_binned_(33.0, 41.0]") + Q("age_binned_(41.0, 50.0]") + Q("age_binned_(50.0, 90.0]") + Q("workclass_?") + Q("workclass_Federal-gov") + Q("workclass_Local-gov") + Q("workclass_Never-worked") + Q("workclass_Private") + Q("workclass_Self-emp-inc") + Q("workclass_Self-emp-not-inc") + Q("workclass_State-gov") + Q("workclass_Without-pay") + Q("fnlwgt_binned_(12284.999, 106648.0]") + Q("fnlwgt_binned_(106648.0, 158662.0]") + Q("fnlwgt_binned_(158662.0, 196338.0]") + Q("fnlwgt_binned_(196338.0, 259873.0]") + Q("fnlwgt_binned_(259873.0, 1484705.0]") + Q("education_10th") + Q("education_11th") + Q("education_12th") + Q("education_1st-4th") + Q("education_5th-6th") + Q("education_7th-8th") + Q("education_9th") + Q("education_Assoc-acdm") + Q("education_Assoc-voc") + Q("education_Bachelors") + Q("education_Doctorate") + Q("education_HS-grad") + Q("education_Masters") + Q("education_Preschool") + Q("education_Prof-sc

In [19]:
Y, X = dmatrices(formula, df2, return_type='dataframe')
y = Y['target'].values

#### Set up the classifier

In [20]:
from sklearn import naive_bayes
model = naive_bayes.MultinomialNB()

#### Fit the model

In [21]:
model.fit(X, y)

MultinomialNB()

#### Test on some training data

In [22]:
print('Prediction')
print(model.predict(X[:10]))
print('Actual')
print(y[:10])

Prediction
[0. 1. 0. 0. 1. 1. 0. 1. 1. 1.]
Actual
[0. 0. 0. 0. 0. 0. 0. 1. 1. 1.]


In [23]:
from sklearn import metrics

prediction_train = model.predict(X)
print(metrics.accuracy_score(y, prediction_train))

0.8188937686189


#### Test

We load the test data from another CSV file.

In [24]:
df_test = pd.read_csv('Classification_2_data/adult.test',
                      sep=', ',
                      names=names)

#### Create the target

In [25]:
df_test['target'] = 0.0
mask = (df_test['income_band'] == '>50K')
df_test['target'][mask] = 1.0

#### Do the feature binning

In [26]:
# Use the bins we created using from the training set
# Now, we need pd.cut instead of pd.qcut
df_test['age_binned'] = pd.cut(df_test['age'],
                               bins=age_bins,
                               include_lowest=True)
df_test['fnlwgt_binned'] = pd.cut(df_test['fnlwgt'],
                                  bins=fnlwgt_bins,
                                  include_lowest=True)
df_test['education_num_binned'] = pd.cut(df_test['education_num'],
                                         bins=education_num_bins,
                                         include_lowest=True)

In [27]:
# Rest is the same as before
df_test['capital_loss_binned'] = '0'
df_test['capital_loss_binned'][df_test['capital_loss'] > 0] = '>0'

df_test['capital_gain_binned'] = '0'
df_test['capital_gain_binned'][df_test['capital_gain'] > 0] = '>0'

df_test['hours_binned'] = '40'
df_test['hours_binned'][df_test['hours_per_week'] < 40] = '<40'
df_test['hours_binned'][df_test['hours_per_week'] > 40] = '>40'

#### Create the design matrices

In [28]:
df_dummies = pd.get_dummies(df_test[categorical_columns],
                            prefix=categorical_columns,
                            columns=categorical_columns)
dummy_column_names = df_dummies.columns.values
df_test2 = pd.concat([df_test, df_dummies], axis=1)

In [29]:
Y_test, X_test = dmatrices(formula, df_test2, return_type='dataframe')
y_test = Y_test['target'].values

#### Run the test

In [30]:
prediction_test = model.predict(X_test)
print(metrics.accuracy_score(y_test, prediction_test))

0.8220625268718138


* Testing accuracy was similar to training accuracy
    * The accuracy on the training set was 0.819
* This is good; it means we are not overfitting
    * With overfitting, we would have great training accuracy,
    * but the classifier would perform poorly on the test set.

#### Class priors

* Class priors are encoded in the `model` variable, after fitting the model to the training data. 
* The priors are available from `model.class_log_prior_`
    * This is not the fraction of each class in the training data, but rather
    * the _logarithm_ of that.

In [31]:
print('Prior probability for the negative class is',)
print(exp(model.class_log_prior_[0]))
print('Prior probability for the positive class is',)
print(exp(model.class_log_prior_[1]))

Prior probability for the negative class is
0.7591904425539763
Prior probability for the positive class is
0.24080955744602456


This says:

* first class (class 0) has prior probability 0.76
* second class (class 1) has probability 0.24

We can quickly check that this is exactly the fraction of these classes in the training data.

In [32]:
df['target'].value_counts() / len(df)

0.0    0.75919
1.0    0.24081
Name: target, dtype: float64

This also says that we can easily achieve 76% accuracy: just predict negative class always.

Our accuracy of 82% is better than this baseline, but perhaps not by a lot.

#### What are the likelihoods?

There is a special attribute called `model.feature_log_prob_` that is available after fitting the training data.

In [33]:
# Log of the likelihoods for class 1 (positive class) for the first 5 features
model.feature_log_prob_[1][:5]

array([-6.35107198, -4.55837662, -3.87227007, -3.75926107, -3.88363922])

In [34]:
feature_stats = DataFrame({'Positive class':np.exp(model.feature_log_prob_[1]),
                           'Negative class':np.exp(model.feature_log_prob_[0])},
                           index=X.columns.values) # <-- feature names
feature_stats[:5]

Positive class  Negative class
Q("age_binned_(16.999, 26.0]")        0.001745        0.021839
Q("age_binned_(26.0, 33.0]")          0.010479        0.015244
Q("age_binned_(33.0, 41.0]")          0.020811        0.014441
Q("age_binned_(41.0, 50.0]")          0.023301        0.011822
Q("age_binned_(50.0, 90.0]")          0.020576        0.013573

#### Which features are most important?

A feature is important (i.e. very **discriminative**) if it is very _likely_ in one class but very _unlikely_ in the other.

Look at the ratio of $$P(\mbox{feature in positive class})$$ to $$P(\mbox{feature in negative class})$$

$$\mbox{If Ratio } \gg 1 \Rightarrow \mbox{presence of feature implies positive class}$$

$$\mbox{If Ratio } \ll 1 \Rightarrow \mbox{presence of feature implies negative class}$$

In [35]:
feature_stats['Positive/Negative Ratio'] = feature_stats['Positive class']\
                                         / feature_stats['Negative class']
feature_stats[:3]

Positive class  Negative class  \
Q("age_binned_(16.999, 26.0]")        0.001745        0.021839   
Q("age_binned_(26.0, 33.0]")          0.010479        0.015244   
Q("age_binned_(33.0, 41.0]")          0.020811        0.014441   

                                Positive/Negative Ratio  
Q("age_binned_(16.999, 26.0]")                 0.079896  
Q("age_binned_(26.0, 33.0]")                   0.687426  
Q("age_binned_(33.0, 41.0]")                   1.441084

$$\mbox{How can we measure if the ratio is far from 1?}$$.

$$\Rightarrow \log\left(Ratio\right) \mbox{ is very large in magnitude}$$

In [36]:
feature_stats['Importance'] = \
    np.abs(np.log(feature_stats['Positive/Negative Ratio']))
feature_stats[:3]

Positive class  Negative class  \
Q("age_binned_(16.999, 26.0]")        0.001745        0.021839   
Q("age_binned_(26.0, 33.0]")          0.010479        0.015244   
Q("age_binned_(33.0, 41.0]")          0.020811        0.014441   

                                Positive/Negative Ratio  Importance  
Q("age_binned_(16.999, 26.0]")                 0.079896    2.527024  
Q("age_binned_(26.0, 33.0]")                   0.687426    0.374801  
Q("age_binned_(33.0, 41.0]")                   1.441084    0.365396

How can we find the top features?

In [37]:
feature_stats.sort_values(by='Importance', ascending=False)[:10]

Positive class  Negative class  \
Q("occupation_Priv-house-serv")          0.000020        0.000464   
Q("relationship_Own-child")              0.000667        0.015561   
Q("education_Preschool")                 0.000010        0.000162   
Q("age_binned_(16.999, 26.0]")           0.001745        0.021839   
Q("education_Doctorate")                 0.003009        0.000336   
Q("education_Prof-school")               0.004156        0.000479   
Q("relationship_Other-relative")         0.000373        0.002940   
Q("education_1st-4th")                   0.000069        0.000507   
Q("occupation_Other-service")            0.001353        0.009828   
Q("marital_status_Never-married")        0.004823        0.031710   

                                   Positive/Negative Ratio  Importance  
Q("occupation_Priv-house-serv")                   0.042295    3.163088  
Q("relationship_Own-child")                       0.042836    3.150375  
Q("education_Preschool")                          0.060596    2.803533  
Q("age_binned_(16.999, 26.0]")                    0.079896    2.527024  
Q("education_Doctorate")                          8.956927    2.192427  
Q("education_Prof-school")                        8.675401    2.160491  
Q("relationship_Other-relative")                  0.126706    2.065888  
Q("education_1st-4th")                            0.135318    2.000129  
Q("occupation_Other-service")                     0.137649    1.983046  
Q("marital_status_Never-married")                 0.152092    1.883267

$$\mbox{If Ratio } \gg 1 \Rightarrow \mbox{presence of feature implies positive class}$$

* So "education_Doctorate" and "education_Prof-school" are markers of positive class (i.e., >=50K salary)